### ISSUE TITLE
No loss logging when using LoraConfig + SFTTrainer

### ISSUE THREAD

**Issue by LukasStruppek** [huggingface/trl#3694]
### Reproduction

Hi,

when using LoRA from peft together with SFTTrainer, the evaluation loss is not logged correctly. However, this can be manually activated by setting ```trainer.can_return_loss = True```. So there seems to be a wrong default parameter for this setting.

```python
from transformers import AutoModelForCausalLM
from datasets import load_dataset
from trl import SFTConfig, SFTTrainer
from peft import LoraConfig

dataset = load_dataset("stanfordnlp/imdb", split="train[:500]")
dataset = dataset.train_test_split(test_size=0.01)

model = AutoModelForCausalLM.from_pretrained("facebook/opt-350m")

peft_config = LoraConfig(
    r=4,
    lora_alpha=16,
    lora_dropout=0.05,
    bias="none",
    target_modules="all-linear",
)

training_args = SFTConfig(max_length=2048,
                          output_dir="/tmp",
                          eval_strategy='steps',
                          eval_steps=10,
                          eval_on_start=True,
                          report_to='wandb')

trainer = SFTTrainer(
    model,
    train_dataset=dataset['train'],
    eval_dataset=dataset['test'],
    args=training_args,
    peft_config=peft_config, # comment out for the first setting without Lora
)

# Add the following line to enable logging of evaluation loss
# trainer.can_return_loss = True

trainer.train()
```

Outputs when NOT using LoRA:
```
{'eval_loss': 3.513484239578247, 'eval_runtime': 0.557, 'eval_samples_per_second': 8.977, 'eval_steps_per_second': 1.795, 'eval_num_tokens': 0.0, 'eval_mean_token_accuracy': 0.3457943797111511, 'epoch': 0}                                                                                                                                                             
{'loss': 3.4785, 'grad_norm': 5.905367851257324, 'learning_rate': 1.903225806451613e-05, 'num_tokens': 24041.0, 'mean_token_accuracy': 0.3565803587436676, 'epoch': 0.16}                                                                                                                                                                                                 
```


Outputs when using LoRA:
```
{'eval_runtime': 0.3156, 'eval_samples_per_second': 15.841, 'eval_steps_per_second': 3.168, 'epoch': 0}                                                                                                                                                                                                                                                                   
{'loss': 3.4884, 'grad_norm': 0.7195760011672974, 'learning_rate': 1.903225806451613e-05, 'num_tokens': 24104.0, 'mean_token_accuracy': 0.35992352962493895, 'epoch': 0.16}                                                                                                                                                                                               
```


Outputs when using LoRA + setting trainer.can_return_loss = True :
```
{'eval_loss': 3.567966938018799, 'eval_runtime': 0.2626, 'eval_samples_per_second': 19.043, 'eval_steps_per_second': 3.809, 'eval_num_tokens': 0.0, 'eval_mean_token_accuracy': 0.3465627133846283, 'epoch': 0}                                                                                                                                                           
{'loss': 3.5367, 'grad_norm': 0.838955819606781, 'learning_rate': 1.903225806451613e-05, 'num_tokens': 21051.0, 'mean_token_accuracy': 0.35161270797252653, 'epoch': 0.16}                                                                                                                                                                                                
```


### System Info

- Platform: Linux-5.4.0-198-generic-x86_64-with-glibc2.35
- Python version: 3.11.10
- TRL version: 0.18.1
- PyTorch version: 2.7.0
- accelerator(s): NVIDIA A100-SXM4-80GB, NVIDIA A100-SXM4-80GB, NVIDIA A100-SXM4-80GB, NVIDIA A100-SXM4-80GB, NVIDIA A100-SXM4-80GB, NVIDIA A100-SXM4-80GB, NVIDIA A100-SXM4-80GB, NVIDIA A100-SXM4-80GB
- Transformers version: 4.52.4
- Accelerate version: 1.3.0
- Accelerate config: not found
- Datasets version: 3.6.0
- HF Hub version: 0.33.0
- bitsandbytes version: 0.46.0
- DeepSpeed version: 0.17.0
- Diffusers version: not installed
- Liger-Kernel version: not installed
- LLM-Blender version: not installed
- OpenAI version: 1.91.0
- PEFT version: not installed
- vLLM version: 0.9.0.1

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by wa008** [comment]
By default, model class with full fine-tune is `OPTForCausalLM ` and model class with lora is [PeftModel](https://github.com/huggingface/peft/blob/d76f3fe98c4a569fc9195cafdc9c0c44921ea4b2/src/peft/mapping_func.py#L115),  there isn't `labels` in forward function of `PeftModel`, `labels` is a sign indicate whether compute and output loss or not in default config(we cannot compute loss without labels), that's why full fine-tune can output eval_loss by default and lora cannot. 

`can_return_loss` is proper parameter to output loss, I think maintaining current logic is acceptable. 


related link
- [find_label_name](https://github.com/huggingface/transformers/blob/b283d52f7f89d9cf3c77cfef233c4cbf700959ff/src/transformers/trainer.py#L786)
- [has_labels](https://github.com/huggingface/transformers/blob/b283d52f7f89d9cf3c77cfef233c4cbf700959ff/src/transformers/trainer.py#L4563)
- [output loss or not](https://github.com/huggingface/transformers/blob/b283d52f7f89d9cf3c77cfef233c4cbf700959ff/src/transformers/trainer.py#L4590)

**Comment by LukasStruppek** [comment]
I am not sure whether I agree that the current setting is reasonable. Given that all model fine-tuning provides some kind of loss and PeftModels are designed to be somehow trained, providing default loss logging would be helpful for many users.

Otherwise, I would suggest adding a parameter to manually activate the loss logging during training and evaluation. Currently, setting the parameter ```can_return_loss``` manually for the object seems a bit hacky and is not well document. Having instead some parameter one can set would be more straight forward.

**Comment by wa008** [comment]
I agree what you mean

This is a known problem of [PeftModel](https://github.com/huggingface/transformers/blob/bbca9782ca1b8b358cc832a1b821aa1b450850da/src/transformers/trainer.py#L781) and trigger warning by default, below is a better(less tricky) option to output eval_loss, hope helpful

```
training_args = SFTConfig(
    label_names = ["labels"], 
    ...
)
```

**Comment by qgallouedec** [comment]
thanks for this very clean issue. This might solve this issue: https://github.com/huggingface/transformers/pull/39265

**Comment by qgallouedec** [comment]
In the meantime, and if it doesn't get merged for some reason, then https://github.com/huggingface/trl/issues/3694#issuecomment-3047100092 from @wa008 is definitely the way to way to go.

**Comment by LukasStruppek** [comment]
Thank you for your support. I think with the PR or the label_name trick, the issue will be resolved.
